File: Forge/src/main/java/hantonik/atomic/core/block/entity/IAtomicTickableBlockEntity.java
package hantonik.atomic.core.block.entity;

import net.minecraft.world.level.block.entity.BlockEntity;
import net.minecraft.world.level.block.entity.BlockEntityTicker;
import net.minecraft.world.level.block.entity.BlockEntityType;

public interface IAtomicTickableBlockEntity {
    void tick();

    static <A extends BlockEntity> BlockEntityTicker<A> createTicker(BlockEntityType<A> type, BlockEntityType<?> expectedType, Class<?> clazz) {
        if (type == expectedType)
            if (IAtomicTickableBlockEntity.class.isAssignableFrom(clazz))
                return (level, pos, state, entity) -> ((IAtomicTickableBlockEntity) entity).tick();

        return null;
    }
}

File: Forge/src/main/kotlin/hantonik/atomic/core/block/entity/IAtomicTickableBlockEntity.kt
package hantonik.atomic.core.block.entity

import net.minecraft.world.level.block.entity.BlockEntity
import net.minecraft.world.level.block.entity.BlockEntityTicker
import net.minecraft.world.level.block.entity.BlockEntityType

interface IAtomicTickableBlockEntity {
    fun tick()

    companion object {
        fun <A : BlockEntity> createTicker(type: BlockEntityType<A>, expectedType: BlockEntityType<*>, clazz: Class<*>): BlockEntityTicker<A>? {
            if (type == expectedType)
                if (IAtomicTickableBlockEntity::class.java.isAssignableFrom(clazz))
                    return BlockEntityTicker { _, _, _, entity -> (entity as IAtomicTickableBlockEntity).tick() }

            return null
        }
    }
}